src/unix/darwin-battery.c

```diff
@@ -0,0 +1,66 @@
+/* Copyright libuv project contributors. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+
+#include <dlfcn.h>
+
+#if !TARGET_OS_IPHONE
+#include "darwin-stub.h"
+#endif
+
+int uv_battery_info(uv_battery_info_t* info) {
+  CFTimeInterval (*pIOPSGetTimeRemainingEstimate)(void);
+  void* iokit_framekwork_handle;
+  int err;
+
+  err = UV_ENOENT;
+  iokit_framekwork_handle = dlopen("/System/Library/Frameworks/"
+                                  "IOKit.framework/IOKit",
+                                  RTLD_LAZY | RTLD_LOCAL);
+
+  if (iokit_framekwork_handle == NULL)
+    goto out;
+
+  *(void **)(&pIOPSGetTimeRemainingEstimate) =
+      dlsym(iokit_framekwork_handle, "IOPSGetTimeRemainingEstimate");
+
+  if (pIOPSGetTimeRemainingEstimate == NULL)
+    goto out;
+
+  err = UV_EINVAL;
+  info->charging = pIOPSGetTimeRemainingEstimate();
```

Comment:
You should assign this to a local variable first because implementation details like the -1 and -2 status codes shouldn't leak out to callers.

Reply:
On it!

How are we going to handle -1 and -2 states? As long as the struct will store only positive numbers?

Reply:
I've added further members to the structure, but I think the question remains: how to handle platform-specific values like -1, or BYTE(255) in Win32 case.

https://github.com/libuv/libuv/pull/4271